mineral-ui/mineral-ui
# [Table]: Add columnKey and rowIndex to table rows and cells

### Description

Provides `columnKey` and `rowIndex` to table cells.  Gives user a way to access other data in row, e.g. previous and next cells.

### Motivation and context

This feature has been requested by users for use with custom cell renderers.

Is it sufficient to provide `rowIndex` and `columnKey` or do we need to provide `rowData`?  Here, we provide the indexes instead of a data object in order to attempt to avoid unnecessary cell re-renders when any row data is changed.   With the current solution, there is no danger of unnecessary re-rendering and the `rowIndex` and `columnKey` can be used to look up information in the main data source.

### Screenshots, videos, or demo, if appropriate

https://table-row-index--mineral-ui.netlify.com/components/table/

### How to test

1. Go to the [custom cell](https://table-row-index--mineral-ui.netlify.com/components/table/custom-cell) example.
2. Add the following block to the render method `CustomCell` class.
```js
const { columnKey: currentKey, rowIndex } = this.props;
const columnKeys = columns.map(column => column.key);
const currentIndex = columnKeys.indexOf(currentKey);
const currentValue = data[rowIndex][currentKey];
const prevKey = columnKeys[currentIndex - 1];
const prevValue = data[rowIndex][prevKey];
const nextKey = columnKeys[currentIndex + 1];
const nextValue = data[rowIndex][nextKey];
console.log({ prevValue, currentValue, nextValue });   
```
3. Note that you can use the new props to access the previous and next data in the row.

### Types of changes

- New feature (non-breaking change which adds functionality)

### Checklist

* [x] Automated tests written and passing - **existing snapshots**
